Allow ifIndex swaps during port discovery

eskyuu opened the pull request:
Add some code on the port discovery module to swap the ifIndex entries of ports

I have noticed that if the SNMP ifIndex for 2 interfaces swap places, and LibreNMS is not using ifIndex for mapping interfaces to ports, the poller can get stuck being unable to update these ports because both ports get a unique constraint violation.  This PR adds some code into the discovery module that will do one of 2 things to fix this:
 - It will swap the ifIndex IDs of the 2 ports within a transaction if both ports are detected in the DB
 - It will immediately update the ifIndex of a port if a new port will need the ifIndex
 - It will delete any ports with the deleted flag if a new port will need the ifIndex

There are probably some conditions that will take 2 discovery cycles to resolve (if a port has not been detected as deleted yet), but the main goal is still to try and preserve the port_id, so it is better to take 2 cycles to re-use a ifIndex than to delete a port that has just had its ifIndex changed.

DO NOT DELETE THE UNDERLYING TEXT

#### Please note

> Please read this information carefully. You can run `./lnms dev:check` to check your code before submitting.

- [ ] Have you followed our [code guidelines?](https://docs.librenms.org/Developing/Code-Guidelines/)
- [ ] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.
- [ ] If my Pull Request makes discovery/polling/yaml changes, I have added/updated [test data](https://docs.librenms.org/Developing/os/Test-Units/).

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

eskyuu committed:
Allow ifIndex swaps during port discovery

 - Add some code on the port discovery module to swap the ifIndex entries of ports

eskyuu committed:
StyleCI

eskyuu committed:
Remove redundant call to array_key_exist() and print_r()

murrant commented on includes/discovery/ports.inc.php:
```diff
@@ -108,6 +109,64 @@
+
+    // Go through remaining conflicts and delete any rows marked as deleted
+    foreach ($ifindex_conflicts as $ifIndex => $port_id) {
+        DB::update('DELETE FROM ports WHERE deleted=1 AND port_id=:port_id', ['port_id' => $port_id]);
```
Ports table only ever soft deletes ports.  We should not be deleting them here.

eskyuu replied:
I am assuming that the constraint exists for a reason, probably because the default configuration uses ifIndex as the port association mode.  Also, my code won't run unless someone changes the port association mode.

I did read an issue where someone was having problems with deleted ports stopping an active port from graphing, and I think that deleting a port is better than having an active port not graphing.

Alternate options I can think of are:
 - Take over the deleted port based on ifIndex instead of deleting it, but this could cause old unrelated data to appear on the interface graph
 - Remove the unique key, but I think this could cause issues with the default port association mode
 - Create a new unique key field which we update to match the value from the port association mode (and add something to make sure this field is synced to the correct value at the start of discovery in case the association mode changes)

If you think showing old data is better than deleting the row, I am happy to have a go at implementing something to do this instead.

murrant replied:
"I am assuming that the constraint exists for a reason" That is a big assumption :D

murrant replied:
I suspect the ifIndex constraint is a pre-fork relic.  Likely predates port association mode.

eskyuu replied:
> The goal should be to keep the port_id mapped to the same port. If a transient condition makes all ports be removed, we need to re-attach the ports when they are back.

I believe that this new code will still work correctly if all ports disappear and then re-appear, if ports all get new ifIndexes.  The only issue is if a small number of ports disappear, and one of them holds onto an ifIndex that has been re-assigned.  In this last scenario, it will still take 2 discovery cycles to delete a port (the first cycle will mark the port as deleted, then the second will delete the port).

I think that this is an acceptable compromise, especially since it's not modifying the behaviour of the system unless someone changes the port association mode.

Having said this, I have suggested 3 options above that I'm willing to implement if we really don't want to delete ports.

laf replied:
> I suspect the ifIndex constraint is a pre-fork relic. Likely predates port association mode.

https://github.com/librenms/librenms/commit/e4a8e284c0c1fb1bcdec070167c8a2e76363c52e#diff-cf16dc0d30091a10c88e72a4d6863d105aee0264d4f8769b0998270551bdb59bR786

It's in from the first commit so yes, existed before the fork so way before port association mode.

murrant commented:
Would removing the ifIndex constraint be easier? or do you think that is protecting us from some bad code?

eskyuu commented:
Having had time to think, are you happy if I just change this to remove the constraint?  This option fixes the problem I'm having.

f0o commented:
I've recently ran into the same issue and simply removed the constraint on my installation via a migration - it seems to work just fine without that constraint.

eskyuu committed:
Adjust to remove ifindex unique constraint

eskyuu committed:
Remove trailing whitespace

eskyuu commented:
Updated to remove the unique constraint.

I have left some logging output changes in place that I think make sense (move the verbose "Adding" message to be a d_echo, added a "+" as the normal output for every port added, and removed a "X" output that was happening for every port).

eskyuu committed:
Update DB schema

laf reviewed (approved):
LGTM
